`testing.numpy_cupy_allclose` does not support test skips

https://github.com/cupy/cupy/pull/2483#discussion_r327902360

>```
>================================================== FAILURES ===================================================
>_______________________ TestConvolveAndCorrelate_param_217.test_convolve_and_correlate ________________________
>/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/parameterized.py:60: in wrap
>    six.raise_from(e_new.with_traceback(e.__traceback__), None)
>/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/parameterized.py:43: in wrap
>    return method(*args, **kwargs)
>/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/helper.py:156: in test_func
>    accept_error=accept_error)
>/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/helper.py:115: in _check_cupy_numpy_error
>    self.fail(msg)
>E   AssertionError: Parameterized test failed.
>E   
>E   Base test method: TestConvolveAndCorrelate.test_convolve_and_correlate
>E   Test parameters:
>E     adtype: <class 'numpy.int32'>
>E     cval: 0.0
>E     filter: convolve
>E     ksize: 3
>E     mode: reflect
>E     origin: 0
>E     output: None
>E     shape: (3, 4)
>E     wdtype: <class 'numpy.int32'>
>E   
>E   AssertionError: Both cupy and numpy raise exceptions
>E   
>E   cupy
>E   Traceback (most recent call last):
>E     File "/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/helper.py", line 28, in _call_func
>E       result = impl(self, *args, **kw)
>E     File "/home/anaruse/work/CUPY/cupy/tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py", line 46, in test_convolve_and_correlate
>E       raise unittest.SkipTest()
>E   unittest.case.SkipTest
>E   
>E   numpy
>E   Traceback (most recent call last):
>E     File "/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/helper.py", line 28, in _call_func
>E       result = impl(self, *args, **kw)
>E     File "/home/anaruse/work/CUPY/cupy/tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py", line 46, in test_convolve_and_correlate
>E       raise unittest.SkipTest()
>E   unittest.case.SkipTest
>```

--- a/cupy/testing/helper.py
+++ b/cupy/testing/helper.py
@@ -80,6 +80,14 @@ def _check_numpy_cupy_error_compatible(cupy_error, numpy_error):
 
 def _check_cupy_numpy_error(self, cupy_error, cupy_tb, numpy_error,
                             numpy_tb, accept_error=False):
+    # Skip the test if both raised SkipTest.
+    if (isinstance(cupy_error, unittest.SkipTest)
+            and isinstance(numpy_error, unittest.SkipTest)):
+        if cupy_error.args != numpy_error.args:
+            raise AssertionError(
+                'Both numpy and cupy were skipped but with different causes.')
+        raise numpy_error  # reraise SkipTest
+
     # For backward compatibility
     if accept_error is True:
         accept_error = Exception

--- a/tests/cupy_tests/testing_tests/test_helper.py
+++ b/tests/cupy_tests/testing_tests/test_helper.py
@@ -2,6 +2,7 @@ import re
 import unittest
 
 import numpy
+import pytest
 import six
 
 import cupy
@@ -423,3 +424,81 @@ class TestShapedRandom(unittest.TestCase):
         self.assertTrue(self.xp.all(0 <= a.imag))
         self.assertTrue(self.xp.all(a.imag < 10))
         self.assertTrue(self.xp.any(a.imag))
+
+
+class TestSkip(unittest.TestCase):
+
+    @testing.numpy_cupy_allclose()
+    def test_allclose(self, xp):
+        raise unittest.SkipTest('Test for skip with @numpy_cupy_allclose')
+        assert False
+
+    @testing.numpy_cupy_array_almost_equal()
+    def test_array_almost_equal(self, xp):
+        raise unittest.SkipTest(
+            'Test for skip with @numpy_cupy_array_almost_equal')
+        assert False
+
+    @testing.numpy_cupy_array_almost_equal_nulp()
+    def test_array_almost_equal_nulp(self, xp):
+        raise unittest.SkipTest(
+            'Test for skip with @numpy_cupy_array_almost_equal_nulp')
+        assert False
+
+    @testing.numpy_cupy_array_max_ulp()
+    def test_array_max_ulp(self, xp):
+        raise unittest.SkipTest('Test for skip with @numpy_cupy_array_max_ulp')
+        assert False
+
+    @testing.numpy_cupy_array_equal()
+    def test_array_equal(self, xp):
+        raise unittest.SkipTest('Test for skip with @numpy_cupy_array_equal')
+        assert False
+
+    @testing.numpy_cupy_array_list_equal()
+    def test_array_list_equal(self, xp):
+        raise unittest.SkipTest(
+            'Test for skip with @numpy_cupy_array_list_equal')
+        assert False
+
+    @testing.numpy_cupy_array_less()
+    def test_less(self, xp):
+        raise unittest.SkipTest('Test for skip with @numpy_cupy_array_less')
+        assert False
+
+    @testing.numpy_cupy_equal()
+    def test_equal(self, xp):
+        raise unittest.SkipTest('Test for skip with @numpy_cupy_equal')
+        assert False
+
+    @testing.numpy_cupy_raises()
+    def test_raises(self, xp):
+        raise unittest.SkipTest('Test for skip with @numpy_cupy_raises')
+        assert False
+
+
+class TestSkipFail(unittest.TestCase):
+
+    @pytest.mark.xfail(strict=True)
+    @testing.numpy_cupy_allclose()
+    def test_different_reason(self, xp):
+        if xp is numpy:
+            raise unittest.SkipTest('skip1')
+        else:
+            raise unittest.SkipTest('skip2')
+
+    @pytest.mark.xfail(strict=True)
+    @testing.numpy_cupy_allclose()
+    def test_only_numpy(self, xp):
+        if xp is numpy:
+            raise unittest.SkipTest('skip')
+        else:
+            return xp.array(True)
+
+    @pytest.mark.xfail(strict=True)
+    @testing.numpy_cupy_allclose()
+    def test_only_cupy(self, xp):
+        if xp is numpy:
+            return xp.array(True)
+        else:
+            raise unittest.SkipTest('skip')

